src/main/scala/mesosphere/marathon/api/v2/PodsResource.scala

```diff
@@ -31,4 +51,218 @@ class PodsResource @Inject() (
+  }
+}
+
+/**
+  * public interfaces, specific to the pods api, that other pieces of the system implement and/or interact with
+  */
+object PodsResource {
+
+  /**
+    * Pod operations that reach this interface have already been authenticated, authorized, and otherwise validated.
+    * TODO(jdef) does it really make sense that deployment ID's are returned an Option[String] instead of String?
+    */
+  trait System {
+
+    /**
+      * Create results in the deployment of a new pod
+      *
+      * @param p     is the new pod to deploy
+      * @param force TODO(jdef) not sure what this means for create
+      * @return the created pod and an optional, stringified deployment ID
+      */
+    def create(p: PodDefinition, force: Boolean): (PodDefinition, Option[String])
+
+    def findAll(s: Selector): Iterable[PodDefinition]
+
+    def find(id: String): PodDefinition
+
+    def update(p: PodDefinition, force: Boolean): (PodDefinition, Option[String])
+
+    def delete(id: String, force: Boolean): (PodDefinition, Option[String])
+
+    // TODO(jdef) leaky! We need a non-API status return type that converts to PodStatus
+    def examine(id: String): PodStatus
+  }
+
+  trait Selector extends Function1[PodDefinition, Boolean]
+
+  object Selector {
+    def apply(f: PodDefinition => Boolean): Selector = new Selector {
+      override def apply(p: PodDefinition): Boolean = f(p)
+    }
+  }
+}
+
+/**
+  * Helpers for internal use (by PodsResource and related tests) only
+  */
+protected object PodsResourceInternal {
+
+  val DEPLOYMENT_ID_HEADER = "MARATHON-Deployment-ID"
+
+  val createPodValidator: Validator[PodDef] = validator[PodDef] { pod =>
+    pod is valid(PodsValidation.podDefValidator)
+    pod.version is empty
+  }
+
+  val updatePodValidator: Validator[PodDef] = validator[PodDef] { pod =>
+    // TODO(jdef) only some fields are mutable, enforce that here
+  }
+
+  case class Config(defaultNetworkName: Option[String])
+
+  object Config {
+    def from(conf: MarathonConf): Config =
+      new Config(defaultNetworkName = conf.defaultNetworkName.get)
+  }
+
+  def withDefaults(podDef: PodDef, config: Config): PodDef =
+    // TODO(jdef) defaults for scaling and scheduling policy should come from RAML (and codegen)
+    podDef.copy(
+      networks = podDef.networks.map { network: Network =>
+        config.defaultNetworkName.collect {
+          case (defaultName: String) if defaultName.nonEmpty && network.name.isEmpty =>
+            network.copy(name = Some(defaultName))
+        }.getOrElse(network)
+      }
+    )
+
+  def decodeBytes(data: Array[Byte], req: HttpServletRequest): String = {
+    val maybeEncoding = Option(req.getCharacterEncoding())
+    val charset = maybeEncoding match {
+      case Some(charsetName) =>
+        java.nio.charset.Charset.forName(charsetName)
+      case None =>
+        StandardCharsets.UTF_8 // TODO(jdef) should this be configurable somewhere?
+    }
+    new String(data, charset)
+  }
```

Comment:
AppsResource uses play for json parsing, which appears to rely on
auto-detection of UTF variants based on the contents of the byte array
(doesn't consider request headers at all)

On Thu, Sep 1, 2016 at 8:58 AM, Karsten Jeschkies <notifications@github.com>
wrote:

> In src/main/scala/mesosphere/marathon/api/v2/PodsResource.scala
> <https://github.com/mesosphere/marathon/pull/4277#discussion_r77167889>:
>
> > +          case (defaultName: String) if defaultName.nonEmpty && network.name.isEmpty =>
> > +            network.copy(name = Some(defaultName))
> > +        }.getOrElse(network)
> > +      }
> > +    )
> > +
> > +  def decodeBytes(data: Array[Byte], req: HttpServletRequest): String = {
> > +    val maybeEncoding = Option(req.getCharacterEncoding())
> > +    val charset = maybeEncoding match {
> > +      case Some(charsetName) =>
> > +        java.nio.charset.Charset.forName(charsetName)
> > +      case None =>
> > +        StandardCharsets.UTF_8 // TODO(jdef) should this be configurable somewhere?
> > +    }
> > +    new String(data, charset)
> > +  }
>
> I'd say so. I would not be surprised if there's not already a similar
> method somewhere :).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mesosphere/marathon/pull/4277/files/06fc49e1c1e9195773710c6777df34dfb1666c41#r77167889>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACPVLM0mtFjvXY1LSoH_3fKLRExkgPu5ks5qlsvtgaJpZM4JxFld>
> .
>